src/backend/distributed/worker/worker_data_fetch_protocol.c

```diff
@@ -1244,3 +1294,100 @@ check_log_statement(List *statementList)
 
 	return false;
 }
+
+
+/*
+ * AlterSequenceMinMax arranges the min and max value of the given sequence. The function
+ * creates ALTER SEQUENCE statemenet which sets the start, minvalue and maxvalue of
+ * the given sequence.
+ *
+ * The function provides the uniqueness by shifting the start of the sequence by
+ * GetLocalGroupId() << 48 + 1 and sets a maxvalue which stops it from passing out any
+ * values greater than: (GetLocalGroupID() + 1) << 48.
+ *
+ * This is to ensure every group of workers passes out values from a unique range,
+ * and therefore that all values generated for the sequence are globally unique.
+ */
+static void
+AlterSequenceMinMax(Oid sequenceId, char *schemaName, char *sequenceName)
+{
+	Datum sequenceNextValueDatum = 0;
+	int64 sequenceNextvalue = 0;
+	Datum sequenceRelationIdDatum = 0;
+	StringInfo startNumericString = makeStringInfo();
+	StringInfo maxNumericString = makeStringInfo();
+	int64 startValue = 0;
+	int64 maxValue = 0;
+	Node *startFloatArg = NULL;
+	Node *maxFloatArg = NULL;
+	AlterSeqStmt *alterSequenceStatement = makeNode(AlterSeqStmt);
+	const char *dummyString = "-";
+
+	sequenceRelationIdDatum = ObjectIdGetDatum(sequenceId);
+	sequenceNextValueDatum = DirectFunctionCall1(nextval_oid,
+												 sequenceRelationIdDatum);
+	sequenceNextvalue = DatumGetInt64(sequenceNextValueDatum);
+
+	startValue = (((int64) GetLocalGroupId()) << 48) + 1;
+	maxValue = startValue + ((int64) 1 << 48);
+
+	alterSequenceStatement->sequence = makeRangeVar(schemaName, sequenceName, -1);
+
+	/*
+	 * DefElem->arg can only hold literal ints up to int4, in order to represent
+	 * larger numbers we need to construct a float represented as a string.
+	 */
+
+	appendStringInfo(startNumericString, "%lu", startValue);
+	startFloatArg = (Node *) makeFloat(startNumericString->data);
+
+	appendStringInfo(maxNumericString, "%lu", maxValue);
+	maxFloatArg = (Node *) makeFloat(maxNumericString->data);
+
+	SetDefElemArg(alterSequenceStatement, "start", startFloatArg);
+	SetDefElemArg(alterSequenceStatement, "minvalue", startFloatArg);
+	SetDefElemArg(alterSequenceStatement, "maxvalue", maxFloatArg);
+
+	/*
+	 * Restart the sequence if the current value of the sequence is outside of the newly
+	 * set min and max values.
+	 */
+	if (sequenceNextvalue < startValue || sequenceNextvalue > maxValue)
```

Comment:
This feels a little wonky. Especially because this function unexpectedly draws from the sequence. What if we just never restart? Group ID should never change without recreating the sequence.